File: backend/src/main/kotlin/com/liyaqa/platform/tenant/model/Tenant.kt
package com.liyaqa.platform.tenant.model

import com.liyaqa.shared.domain.OrganizationLevelEntity
import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.EnumType
import jakarta.persistence.Enumerated
import jakarta.persistence.Table
import java.time.Instant
import java.util.UUID

@Entity
@Table(name = "tenants")
class Tenant(
    id: UUID = UUID.randomUUID(),

    @Column(name = "facility_name", nullable = false)
    var facilityName: String,

    @Column(name = "facility_name_ar")
    var facilityNameAr: String? = null,

    @Column(name = "subdomain", unique = true)
    var subdomain: String? = null,

    @Column(name = "cr_number")
    var crNumber: String? = null,

    @Column(name = "vat_number")
    var vatNumber: String? = null,

    @Column(name = "contact_email", nullable = false)
    var contactEmail: String,

    @Column(name = "contact_phone")
    var contactPhone: String? = null,

    @Column(name = "address")
    var address: String? = null,

    @Column(name = "city")
    var city: String? = null,

    @Column(name = "region")
    var region: String? = null,

    @Column(name = "country", nullable = false)
    var country: String = "SA",

    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false)
    var status: TenantStatus = TenantStatus.PROVISIONING,

    @Column(name = "subscription_plan_id")
    var subscriptionPlanId: UUID? = null,

    @Column(name = "deal_id")
    var dealId: UUID? = null,

    @Column(name = "organization_id")
    var organizationId: UUID? = null,

    @Column(name = "club_id")
    var clubId: UUID? = null,

    @Column(name = "onboarded_by")
    var onboardedBy: UUID? = null,

    @Column(name = "onboarded_at")
    var onboardedAt: Instant? = null,

    @Column(name = "deactivated_at")
    var deactivatedAt: Instant? = null,

    @Column(name = "data_retention_until")
    var dataRetentionUntil: Instant? = null,

    @Column(name = "metadata", columnDefinition = "TEXT")
    var metadata: String? = null

) : OrganizationLevelEntity(id) {

    fun changeStatus(newStatus: TenantStatus) {
        val allowed = VALID_TRANSITIONS[status]
            ?: throw IllegalStateException("No transitions defined for status: $status")
        require(newStatus in allowed) {
            "Invalid transition from $status to $newStatus. Allowed: $allowed"
        }
        status = newStatus
        if (newStatus == TenantStatus.ACTIVE) {
            onboardedAt = Instant.now()
        }
        if (newStatus == TenantStatus.DEACTIVATED) {
            deactivatedAt = Instant.now()
        }
    }

    companion object {
        val VALID_TRANSITIONS: Map<TenantStatus, Set<TenantStatus>> = mapOf(
            TenantStatus.PROVISIONING to setOf(TenantStatus.ACTIVE, TenantStatus.DEACTIVATED),
            TenantStatus.ACTIVE to setOf(TenantStatus.SUSPENDED, TenantStatus.DEACTIVATED),
            TenantStatus.SUSPENDED to setOf(TenantStatus.ACTIVE, TenantStatus.DEACTIVATED),
            TenantStatus.DEACTIVATED to setOf(TenantStatus.ARCHIVED),
            TenantStatus.ARCHIVED to emptySet()
        )

        fun create(
            facilityName: String,
            contactEmail: String,
            onboardedBy: UUID? = null,
            dealId: UUID? = null
        ): Tenant {
            return Tenant(
                facilityName = facilityName,
                contactEmail = contactEmail,
                onboardedBy = onboardedBy,
                dealId = dealId
            )
        }
    }
}

File: frontend/shared/src/types/platform/tenant.ts
import type { UUID } from "../api";

/**
 * Tenant status
 */
export type TenantStatus = "PROVISIONING" | "ACTIVE" | "SUSPENDED" | "DEACTIVATED" | "ARCHIVED";

/**
 * Full tenant response — aligned with backend TenantResponse
 */
export interface TenantResponse {
  id: UUID;
  facilityName: string;
  facilityNameAr?: string;
  subdomain?: string;
  crNumber?: string;
  vatNumber?: string;
  contactEmail: string;
  contactPhone?: string;
  address?: string;
  city?: string;
  region?: string;
  country: string;
  status: TenantStatus;
  subscriptionPlanId?: UUID;
  dealId?: UUID;
  organizationId?: UUID;
  clubId?: UUID;
  onboardedBy?: UUID;
  onboardedAt?: string;
  deactivatedAt?: string;
  dataRetentionUntil?: string;
  metadata?: string;
  createdAt: string;
  updatedAt: string;
}

/**
 * Tenant summary for listings — aligned with backend TenantSummaryResponse
 */
export interface TenantSummaryResponse {
  id: UUID;
  facilityName: string;
  status: TenantStatus;
  contactEmail: string;
  createdAt: string;
}

/**
 * Onboarding checklist item
 */
export interface OnboardingChecklistItem {
  step: string;
  completed: boolean;
  completedAt?: string;
  completedBy?: UUID;
  notes?: string;
}

/**
 * Onboarding checklist response (progress summary from backend)
 */
export interface OnboardingChecklistResponse {
  totalSteps: number;
  completedSteps: number;
  percentage: number;
  items: OnboardingChecklistItem[];
}

/**
 * Data export job
 */
export interface DataExportJobResponse {
  id: UUID;
  tenantId: UUID;
  status: "PENDING" | "IN_PROGRESS" | "COMPLETED" | "FAILED";
  format: "JSON" | "CSV";
  requestedBy: UUID;
  startedAt?: string;
  completedAt?: string;
  fileUrl?: string;
  fileSizeBytes?: number;
  errorMessage?: string;
  expiresAt?: string;
  createdAt: string;
}

/**
 * Deactivation log entry
 */
export interface DeactivationLogResponse {
  id: UUID;
  tenantId: UUID;
  reason: string;
  notes?: string;
  deactivatedBy: UUID;
  previousStatus: TenantStatus;
  createdAt: string;
}

/**
 * Tenant filter parameters
 */
export interface TenantFilters {
  status?: TenantStatus;
  search?: string;
  page?: number;
  size?: number;
  sortBy?: string;
  sortDirection?: "asc" | "desc";
}

/**
 * Provision tenant request — aligned with backend ProvisionTenantRequest
 */
export interface ProvisionTenantRequest {
  facilityName: string;
  facilityNameAr?: string;
  subdomain?: string;
  crNumber?: string;
  vatNumber?: string;
  contactEmail: string;
  contactPhone?: string;
  address?: string;
  city?: string;
  region?: string;
  country?: string;
  subscriptionPlanId?: UUID;
  metadata?: string;
}

/**
 * Update tenant request — aligned with backend UpdateTenantRequest
 */
export interface UpdateTenantRequest {
  facilityName?: string;
  facilityNameAr?: string;
  subdomain?: string;
  crNumber?: string;
  vatNumber?: string;
  contactEmail?: string;
  contactPhone?: string;
  address?: string;
  city?: string;
  region?: string;
  country?: string;
  subscriptionPlanId?: UUID;
  metadata?: string;
}

/**
 * Tenant status config
 */
export const TENANT_STATUS_CONFIG: Record<
  TenantStatus,
  { labelEn: string; labelAr: string; color: string; bgColor: string }
> = {
  PROVISIONING: {
    labelEn: "Provisioning",
    labelAr: "جاري التجهيز",
    color: "text-blue-600 dark:text-blue-400",
    bgColor: "bg-blue-50 dark:bg-blue-950/30",
  },
  ACTIVE: {
    labelEn: "Active",
    labelAr: "نشط",
    color: "text-green-600 dark:text-green-400",
    bgColor: "bg-green-50 dark:bg-green-950/30",
  },
  SUSPENDED: {
    labelEn: "Suspended",
    labelAr: "معلق",
    color: "text-yellow-600 dark:text-yellow-400",
    bgColor: "bg-yellow-50 dark:bg-yellow-950/30",
  },
  DEACTIVATED: {
    labelEn: "Deactivated",
    labelAr: "معطل",
    color: "text-red-600 dark:text-red-400",
    bgColor: "bg-red-50 dark:bg-red-950/30",
  },
  ARCHIVED: {
    labelEn: "Archived",
    labelAr: "مؤرشف",
    color: "text-gray-600 dark:text-gray-400",
    bgColor: "bg-gray-50 dark:bg-gray-950/30",
  },
};
